README.md
# tree_transversal

As part of an Assignment for Artifical Intelligence, the task was to create to read a txt file, create a tree consisting of nodes and return in Breadth-First and Depth-First order. 

The program consists of 2 classes; one to create nodes and the other one to create the tree. The Tree class will extract the data from the txt file and for each 'number', the Node class will instantiate objects. These objects will hold information such as 'parents','childs' the node will have. 

Under the Tree class, it will create the actual tree and return nodes in BFS/DFS order.




CHANGE_LOG
HPEnd 7.3

* Added recursion for BFS

********************************************************
Older version

HPEnd 7.2

* added ability to transverse a tree using DFS
* minor big fixes

Version 7 - HpEnd

* re-designed and re-invented version 5
* objects are no longer instantiated using lists
* self.obj removed
* added new feature - Auto Root; Automatically finds the root
* new algorithm implemented; objects are instantiated, assigned and sorted
using a single loop
* removed lists (ID,tmp)
* removed unecessary loops(IDs,inst)
* get_content and set_content no longer used
* major bug fixes



Version 5.1 - RollBX

* rollback to version 5

Version 6.2

* temporary bug changed to permanently fixed
* minor pointer fixes


Version 6.1

* Bug fixed (temporary)
* pointer added (set_content -> self.id)
* get_all_nodes_<method> now returns id node with content


Version 6 - ADvd

* New algorithm implemented
* Added ability to transverse a tree with unlimited nodes
* set_content and get_content fixed
* bf_nodes method now uses get_content
* Added ability to read an unordered tree and automatically sort it
* Create objects dynamically

Version 5

* Implemented "linked list" style tree tranversal
* instantiation happens using lists

Version 4 - Majrx

* skips version 3 due to major updates
* nodes are now instantiated dynamically accessing a list
* method to sort using BFS implemented
* tree is built using BFS
* new effecient way to find for ids and edges
* new logical method to recognize parents and edges from txt files


Version 2

* fully new algorithm; 2 classes implemented
* Node class gets ID and nodes from txt file and uses the methods
* Tree class reads txt file, builds and sorts the tree
* Nodes created manually



Version 1 - Initial release

* Initial algorithm reads the file, extrats the ids and edges from it
* Uses BFS and returns the nodes in order


AIv2.py
class Node(object):
    def __init__(self,id):
        """
        Default constructor

        Parameters
        ----------
        id : int
        """
        self.id = id
        self.children = []
        self.parent=None
        
        
    def set_parent(self,parent):
        self.parent = parent

    def get_parent(self):
        return self.parent

    def add_child(self,child):
        self.children.append(child)

    def get_children(self):
        return self.children

    def set_content(self,content):
        self.content = content

    def get_content(self):
        return self.content

    def __str__(self):
        child_ids = [c.id for c in self.children]
        if self.get_parent()==None:
            ## More info can be useful for testing
            #return "[ROOT {}->{}]".format(self.id,child_ids)
            return "({})".format(self.id)
        else:
            ##return "[{}->{}->{}]".format(self.parent.id,self.id,child_ids)
            return "({})".format(self.id)

    def __repr__(self):
        return self.__str__()
    

class Tree(object):
    
    
    def __init__(self,description_file):
        self.description_file = description_file
        self.make_tree()

    def make_tree(self):
        ids, edges = self.get_ids_and_edges() # this makes it easier but not strictly required
        
        #all nodes existing in the tree created 
        root = Node(0) # root
        node1 = Node(1) 
        node2 = Node(2)
        node3 = Node(3)
        node4 = Node(4)
        node5 = Node(5)
        node6 = Node(6)

        root.add_child(node1) # add left subtree to root # child = 1
        root.add_child(node2) # add right subtree to root child = 1,2


        node1.set_parent(root) # left subtree becomes the parent
        node1.add_child(node3) # add child to left subtree child = 1,2,3
        node1.add_child(node4) # child to left subtree child = 1,2,3,4

        node3.set_parent(node1)
        node4.set_parent(node1)
        
        node2.set_parent(root)
        node2.add_child(node5) # child = 1,2,3,4,5
        node2.add_child(node6) # child = 1,2,3,4,5,6
        
        
        node5.set_parent(node2)
        root.set_content("a")
        node6.set_parent(node2)
        
        print(root)
        
        print(node3)
        print(node4)
        
        print(node5)
        print(node6)
        print(root.children)

        ids1 = [root,node1,node2,node3,node4,node5,node6]
        return ids1
        
    
    def get_all_nodes_BF(self):
        bf_nodes = self.make_tree()
        
            
             
        ## complete the code here; this is required
        ## return the created list of nodes
        return bf_nodes

    def get_all_nodes_DF(self):
        ## fill the list with all nodes in Depth-First order
        df_nodes = []
        ## complete the code here; this is required

        ## return the created list of nodes
        return df_nodes

    def get_ids_and_edges(self):
        """read the tree description file and retrieve the nodes 
        and edges in between of them"""
        f = open(self.description_file,"r")
        ids = [] #nodes
        edges = [] #edges
        tmp = [] # list for temp
        increment = 0

        # loop throught the file
        for line in f: 
            tmp = [line.rstrip("\n")] # add lines to the temporary list
            edges.append(tmp) # add tmp list to the main edge list
            # format :  (edge) -> [parent][child]
            increment += 1
        increment2 = 0
        for x in edges:
            node = x[increment2][0:1]
            node1 = x[increment2][2:3]
            

        ids.append(node)
            
            
         
        print(edges)
        

            
        ## complete the code her

            
        return ids,edges
  
"""This is the code used for testing. 
DO NOT MODIFY.
"""
tree_description = "tree_description_02.txt"
tree = Tree(tree_description)
bf_nodes = tree.get_all_nodes_BF()
print("bf_nodes: {}".format(bf_nodes))

df_nodes = tree.get_all_nodes_DF()
print("df_nodes: {}".format(df_nodes))






"""This partial code is clearly not doing anything but provides the complete code for the Node class and partial code for the Tree class. In the end, your code should return the following:
bf_nodes: [(0), (1), (2), (3), (4), (5), (6)]
df_nodes: [(0), (2), (6), (5), (1), (4), (3)]
I will test the code with another tree description file."""


AIv3.py
class Node(object):
    def __init__(self,id):
        """
        Default constructor

        Parameters
        ----------
        id : int
        """
        self.id = id
        self.children = []
        self.parent=None

    def set_parent(self,parent):
        self.parent = parent

    def get_parent(self):
        return self.parent

    def add_child(self,child):
        self.children.append(child)

    def get_children(self):
        return self.children

    def set_content(content):
        self.content = content

    def get_content(content):
        return self.content

    def __str__(self):
        child_ids = [c.id for c in self.children]
        if self.get_parent()==None:
            ## More info can be useful for testing
            return "[ROOT {}->{}]".format(self.id,child_ids)
            return "({})".format(self.id)
        else:
            return "[{}->{}->{}]".format(self.parent.id,self.id,child_ids)
            return "({})".format(self.id)

    def __repr__(self):
        return self.__str__()
    

class Tree(object):
    
    
    def __init__(self,description_file):
        self.description_file = description_file
        self.make_tree()

    def make_tree(self):
        ids, edges = self.get_ids_and_edges() # this makes it easier but not strictly required
        
        #all nodes existing in the tree created 
        root = Node(ids[0]) # root
        left_sub = Node(ids[1]) 
        right_sub = Node(ids[2])
        node3 = Node(ids[3])
        node4 = Node(ids[4])
        node5 = Node(ids[5])
        node6 = Node(ids[6])

        root.add_child(left_sub) # add left subtree to root # child = 1
        root.add_child(right_sub) # add right subtree to root child = 1,2


        left_sub.set_parent(root) # left subtree becomes the parent
        left_sub.add_child(node3) # add child to left subtree child = 1,2,3
        left_sub.add_child(node4) # child to left subtree child = 1,2,3,4

        node3.set_parent(left_sub)
        node4.set_parent(left_sub)
        
        right_sub.set_parent(root)
        right_sub.add_child(node5) # child = 1,2,3,4,5
        right_sub.add_child(node6) # child = 1,2,3,4,5,6
        
        node5.set_parent(right_sub)
        #none5.add_child(None)
        node6.set_parent(right_sub)
        
##        print(root)
##        print(left_sub)
##        print(node3)
##        print(node4)
##        print(right_sub)
##        print(node5)
##        print(node6)
##        print(root.children)

        ids1 = [root,left_sub,right_sub,node3,node4,node5,node6]
        return ids1
        
    
    def get_all_nodes_BF(self):
        bf_nodes = []
        
            
             
        ## complete the code here; this is required
        ## return the created list of nodes
        #return bf_nodes

    def get_all_nodes_DF(self):
        ## fill the list with all nodes in Depth-First order
        df_nodes = []
        ## complete the code here; this is required

        ## return the created list of nodes
        return df_nodes

    def get_ids_and_edges(self):
        """read the tree description file and retrieve the nodes 
        and edges in between of them"""
        f = open(self.description_file,"r")
        ids = [] #nodes
        edges = [] #edges
        tmp = [] # list for temp
        parents = []
        children = [] # list for chilren
        increment = 0

        # loop throught the file
        for line in f: 
            tmp = [line.rstrip("\n")] # add lines to the temporary list
            edges.append(tmp) # add tmp list to the main edge list
            # format :  (edge) -> [parent][child]
            child = edges[increment][0][2:3] # get the child node in the edge
            parent = int(edges[increment][0][0:1]) # get the parent node
            if parent not in parents: # check for duplicated parent nodes in parent list
                parents.append(int(parent)) # add parent to the list
            children.append(child) # add children to list
            increment += 1
        

        # loop to find childs from parents 
        for x in parents: # iterates every parent to find for child
            if x not in ids:
                ids.append(int(x)) # check if parent is not repeated in ID list
            increment2 = 0
            print("parent",x)
            for y in edges:
                if int(edges[increment2][0][0:1]) == int(x):
                    ids.append(int(edges[increment2][0][2:3])) # ID list in BF order
                    print("Child",edges[increment2][0][2:3])
                increment2 += 1

            
         
##        print(edges)
##        print(ids)

            
        ## complete the code her

            
        return ids,edges
  
"""This is the code used for testing. 
DO NOT MODIFY.
"""
tree_description = "tree_description_01.txt"
tree = Tree(tree_description)
bf_nodes = tree.get_all_nodes_BF()
print("bf_nodes: {}".format(bf_nodes))

df_nodes = tree.get_all_nodes_DF()
print("df_nodes: {}".format(df_nodes))






"""This partial code is clearly not doing anything but provides the complete code for the Node class and partial code for the Tree class. In the end, your code should return the following:
bf_nodes: [(0), (1), (2), (3), (4), (5), (6)]
df_nodes: [(0), (2), (6), (5), (1), (4), (3)]
I will test the code with another tree description file."""


AIv5.py
class Node(object):
    def __init__(self,id):
        """
        Default constructor

        Parameters
        ----------
        id : int
        """
        self.id = id
        self.children = []
        self.parent=None

    def set_parent(self,parent):
        self.parent = parent

    def get_parent(self):
        return self.parent

    def add_child(self,child):
        self.children.append(child)

    def get_children(self):
        return self.children

    def set_content(self,content):
        self.content = content  

    def get_content(self):
        return self.content

    def __str__(self):
        child_ids = [c.id for c in self.children]
        if self.get_parent()==None:
            ## More info can be useful for testing
            #return "[ROOT {}->{}]".format(self.id,child_ids)
            return "({})".format(self.id)
        else:
            #return "[{}->{}->{}]".format(self.parent.id,self.id,child_ids)
            return "({})".format(self.id)

    def __repr__(self):
        return self.__str__()
    

class Tree(object):
    
    
    def __init__(self,description_file):
        self.description_file = description_file
        self.obj = []
        self.make_tree()

    def make_tree(self):
        ids, parents,edges = self.get_ids_and_edges() # this makes it easier but not strictly required

        self.obj = ["a","b","c","d","e","f","g"]
        inc = 0
        for x in self.obj:
            self.obj[inc] = Node(ids[inc])
            #print(self.obj[inc])
            inc += 1
            
        increment = 0
        childinc = 1
        for x in parents: # iterates every parent to find for child
            print("parent",x)
            current = self.obj[increment] # current node
            increment2 = 0
            for y in edges:
                if int(edges[increment2][0][0:1]) == int(x):
                    current.add_child(self.obj[childinc]) # add child to current node
                    self.obj[childinc].set_parent(current) # set current node as parent
                    #print(self.obj[increment])
                    print("Child",edges[increment2][0][2:3])
                    childinc += 1
                increment2 += 1
            increment += 1
    
        
    def get_all_nodes_BF(self):
        bf_nodes = []
        
        for x in self.obj:
            bf_nodes.append(x)
        
            
             
        ## complete the code here; this is required
        ## return the created list of nodes
        return bf_nodes

    def get_all_nodes_DF(self):
        ## fill the list with all nodes in Depth-First order
        df_nodes = []
        ## complete the code here; this is required

        ## return the created list of nodes
        return df_nodes

    def get_ids_and_edges(self):
        """read the tree description file and retrieve the nodes 
        and edges in between of them"""
        f = open(self.description_file)
        ids = [] #nodes
        edges = [] #edges
        tmp = [] # list for temp
        parents = []
        children = [] # list for chilren
        increment = 0

        # loop throught the file
        for line in f: 
            tmp = [line.rstrip("\n")] # add lines to the temporary list
            edges.append(tmp) # add tmp list to the main edge list
            child = edges[increment][0][2:3] # get the child node in the edge
            parent = int(edges[increment][0][0:1]) # get the parent node
            if parent not in parents: # check for duplicated parent nodes in parent list
                parents.append(int(parent)) # add parent to the list
            children.append(child) # add children to list
            increment += 1
        

        # loop to find childs from parents 
        for x in parents: # iterates every parent to find for child
            if x not in ids:
                ids.append(int(x)) # check if parent is not repeated in ID list
            increment2 = 0
            #print("parent",x)
            for y in edges:
                if int(edges[increment2][0][0:1]) == int(x):
                    ids.append(int(edges[increment2][0][2:3])) # ID list in BF order
                    #print("Child",edges[increment2][0][2:3])
                increment2 += 1

            
         
##        print(edges)
##        print(ids)
            
        ## complete the code her

            
        return ids,parents,edges
  
"""This is the code used for testing. 
DO NOT MODIFY.
"""
tree_description = "tree_description_01.txt"
tree = Tree(tree_description)
bf_nodes = tree.get_all_nodes_BF()
print("bf_nodes: {}".format(bf_nodes))

df_nodes = tree.get_all_nodes_DF()
print("df_nodes: {}".format(df_nodes))






"""This partial code is clearly not doing anything but provides the complete code for the Node class and partial code for the Tree class. In the end, your code should return the following:
bf_nodes: [(0), (1), (2), (3), (4), (5), (6)]
df_nodes: [(0), (2), (6), (5), (1), (4), (3)]
I will test the code with another tree description file."""


AIv7.3.py

class Node(object):
    def __init__(self,id):
        """
        Default constructor

        Parameters
        ----------
        id : int
        """
        self.id = id
        self.children = []
        self.parent=None

    def set_parent(self,parent):
        self.parent = parent

    def get_parent(self):
        return self.parent

    def add_child(self,child):
        self.children.append(child)

    def get_children(self):
        return self.children

    def set_content(self,content):
        self.content = content

    def get_content(self):
        return self.content

    def __str__(self):
        child_ids = [c.id for c in self.children]
        if self.get_parent()==None:
            return "[ROOT {}->{}]".format(self.id,child_ids)
            return "({})".format(self.id)
        else:
            return "[{}->{}->{}]".format(self.parent.id,self.id,child_ids)
            return "({})".format(self.id)

    def __repr__(self):
        return self.__str__()
    

class Tree(object):
    
    
    def __init__(self,description_file):
        self.description_file = description_file
        self.bf_nodes = []
        self.parents = []
        self.df_nodes = []
        self.root = None
        self.make_tree()

    def make_tree(self):
        parents,edges = self.get_ids_and_edges()
  

        for x in parents: # iterates every parent to find for child
            increment2 = 0
            parent = Node(x)
            parent.set_parent(self.root)
            for y in edges:
                if int(edges[increment2][0:1]) == int(x):
                    child = Node(edges[increment2][2:4])
                    child.set_parent(parent) # set current node as parent 
                    parent.add_child(child) # add child to current node
                    if parent.get_parent() == None:
                        self.root = parent
 #                   self.bf_nodes.append(child)
                increment2 += 1 
            self.parents.append(child.get_parent())

        
       # self.bf_nodes.insert(0,self.root)
        self.BFS_method(self.parents)

    



        
            
    def BFS_method(self,node):
 
        for x in node:
            if type(x.get_children())==list:
                self.bf_nodes.append(x)
                self.BFS_method(x.get_children())
    
        ## return the created list of nodes
        return self.bf_nodes

    def get_BFS(self):
        
        
        
        return self.bf_nodes

    def get_all_nodes_DF(self):


        self.parents.remove(self.root)# root removed from list
        inc = 0
        for y in reversed(self.parents):
            inc2 = 0
            self.df_nodes.append(y)
            for x in reversed(y.get_children()):
                self.df_nodes.append(x)
                inc2 += 1
            inc += 1
        self.df_nodes.insert(0,self.root) # root added back
        
        ## return the created list of nodes
      
        return self.df_nodes

    def get_ids_and_edges(self):
        """read the tree description file and retrieve the nodes 
        and edges in between of them"""
        f = open(self.description_file)
        edges = [] #edges
        parents = []
        increment = 0
    
        # loop throught the file
        for line in f:  # add lines to the temporary list
            edges.append(line.rstrip("\n")) # add tmp list to the main edge list
            parents.append(int(edges[increment][0:1])) # add parent to the list
            increment += 1

        parents = set(parents)
        return parents,edges

  
"""This is the code used for testing.
"""
tree_description = "tree_description_01.txt"
tree = Tree(tree_description)

bf_nodes = tree.get_BFS()
print("bf_nodes: {}".format(bf_nodes))

df_nodes = tree.get_all_nodes_DF()
print("df_nodes: {}".format(df_nodes))






AIv7.2.py
"""
@SumiteR


Version 7 - HpEnd

* re-designed and re-invented version 5
* objects are no longer instantiated using lists
* self.obj removed
* added new feature - Auto Root; Automatically finds the root
* new algorithm implemented; objects are instantiated, assigned and sorted
using a single loop
* removed lists (ID,tmp)
* removed unecessary loops(IDs,inst)
* get_content and set_content no longer used
* major bug fixes


*********************************************************
Older version
Version 5.1 - RollBX

* rollback to version 5

Version 6.2

* temporary bug changed to permanently fixed
* minor pointer fixes


Version 6.1

* Bug fixed (temporary)
* pointer added (set_content -> self.id)
* get_all_nodes_<method> now returns id node with content


Version 6 - ADvd

* New algorithm implemented
* Added ability to transverse a tree with unlimited nodes
* set_content and get_content fixed
* bf_nodes method now uses get_content
* Added ability to read an unordered tree and automatically sort it
* Create objects dynamically

Version 5

* Implemented "linked list" style tree tranversal
* instantiation happens using lists

Version 4 - Majrx

* skips version 3 due to major updates
* nodes are now instantiated dynamically accessing a list
* method to sort using BFS implemented
* tree is built using BFS
* new effecient way to find for ids and edges
* new logical method to recognize parents and edges from txt files


Version 2

* fully new algorithm; 2 classes implemented
* Node class gets ID and nodes from txt file and uses the methods
* Tree class reads txt file, builds and sorts the tree
* Nodes created manually



Version 1 - Initial release

* Initial algorithm reads the file, extrats the ids and edges from it
* Uses BFS and returns the nodes in order

"""
class Node(object):
    def __init__(self,id):
        """
        Default constructor

        Parameters
        ----------
        id : int
        """
        self.id = id
        self.children = []
        self.parent=None

    def set_parent(self,parent):
        self.parent = parent

    def get_parent(self):
        return self.parent

    def add_child(self,child):
        self.children.append(child)

    def get_children(self):
        return self.children

    def set_content(self,content):
        self.content = content

    def get_content(self):
        return self.content

    def __str__(self):
        child_ids = [c.id for c in self.children]
        if self.get_parent()==None:
            ## More info can be useful for testing
            #return "[ROOT {}->{}]".format(self.id,child_ids)
            return "({})".format(self.id)
        else:
            #return "[{}->{}->{}]".format(self.parent.id,self.id,child_ids)
            return "({})".format(self.id)

    def __repr__(self):
        return self.__str__()
    

class Tree(object):
    
    
    def __init__(self,description_file):
        self.description_file = description_file
        self.bf_nodes = []
        self.parents = []
        self.df_nodes = []
        self.root = None
        self.make_tree()

    def make_tree(self):
        parents,edges = self.get_ids_and_edges() # this makes it easier but not strictly required
  

        for x in parents: # iterates every parent to find for child
            increment2 = 0
            parent = Node(x)
            parent.set_parent(self.root)
            for y in edges:
                if int(edges[increment2][0:1]) == int(x):
                    child = Node(edges[increment2][2:4])
                    child.set_parent(parent) # set current node as parent 
                    parent.add_child(child) # add child to current node
                    if parent.get_parent() == None:
                        self.root = parent
                    self.bf_nodes.append(child)
                increment2 += 1 
            self.parents.append(child.get_parent())

        
        self.bf_nodes.insert(0,self.root)
            
    def get_all_nodes_BF(self):
        """
        make_tree() uses Level order which is similar to the BF search,
        therefore bf_nodes list is appended simultaneous with the loop.
        """
        ## complete the code here; this is required
        ## return the created list of nodes
        return self.bf_nodes

    def get_all_nodes_DF(self):
        ## fill the list with all nodes in Depth-First order
        
        ## complete the code here; this is required

        self.parents.remove(self.root)# root removed from list
        inc = 0
        for y in reversed(self.parents):
            inc2 = 0
            self.df_nodes.append(y)
            for x in reversed(y.get_children()):
                self.df_nodes.append(x)
                inc2 += 1
            inc += 1
        self.df_nodes.insert(0,self.root) # root added back


        
        ## return the created list of nodes
        
        return self.df_nodes

    def get_ids_and_edges(self):
        """read the tree description file and retrieve the nodes 
        and edges in between of them"""
        f = open(self.description_file)
        edges = [] #edges
        parents = []
        increment = 0
        
    ## complete the code here
        # loop throught the file
        for line in f:  # add lines to the temporary list
            edges.append(line.rstrip("\n")) # add tmp list to the main edge list
            parents.append(int(edges[increment][0:1])) # add parent to the list
            increment += 1

        parents = set(parents)
        return parents,edges
  
"""This is the code used for testing. 
DO NOT MODIFY.
"""
tree_description = "tree_description_01.txt"
tree = Tree(tree_description)
bf_nodes = tree.get_all_nodes_BF()
print("bf_nodes: {}".format(bf_nodes))

df_nodes = tree.get_all_nodes_DF()
print("df_nodes: {}".format(df_nodes))






"""This partial code is clearly not doing anything but provides the complete code for the Node class and partial code for the Tree class. In the end, your code should return the following:
bf_nodes: [(0), (1), (2), (3), (4), (5), (6)]
df_nodes: [(0), (2), (6), (5), (1), (4), (3)]
I will test the code with another tree description file."""
